### PR TITLE
External access to etcd via masters ELB

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#1388 and implements #151

This PR adds a second listener to the masters ELB for etcd. It also creates a DNS record in Route 53 for etcd. On delete both the api and etcd records are deleted.

The records look like:
`$component.$customer.aws.giantswarm.io`, i.e. `apiserver.example.aws.giantswarm.io`. That means we added an additional level in the domain hierarchy.

@JosephSalisbury @rossf7 are there any changes in the `certificate-operator` to be made?

We need to expose etcd externally because its used by the worker nodes for calico. Only the client port 2379 is exposed because etcd runs only on the master. Port 2380 will need to be opened later for server to server connections when etcd is made HA. The cloudconfig was changed in giantswarm/k8scloudconfig/pull/58 to only use port 2379.

From the [Calico architecture document](http://docs.projectcalico.org/v2.1/reference/architecture/):
> Calico uses etcd to provide the communication between [its] components and as a consistent data store, which ensures Calico can always build an accurate network.

### CODE REVIEW

File: service/create/loadbalancer.go [hunk 1]
```diff
@@ -65,18 +67,31 @@ func (s *Service) createLoadBalancer(input LoadBalancerInput) error {
...
+
+	etcdRecordSet := &awsresources.RecordSet{
+		Client:       input.Clients.Route53,
+		Resource:     lb,
+		Domain:       input.Cluster.Spec.Cluster.Etcd.Domain,
+		HostedZoneID: hz.ID(),
+	}
+
+	if err := etcdRecordSet.CreateOrFail(); err != nil {
+		return microerror.MaskAny(fmt.Errorf("error registering DNS '%s'", errgo.Details(err)))
```
Comment (xh3b4sd) [inline service/create/loadbalancer.go]: It was there before but I would like to see a typed error being used.

File: service/create/loadbalancer.go [hunk 2]
```diff
@@ -42,7 +44,7 @@ func (s *Service) createLoadBalancer(input LoadBalancerInput) error {
 		s.logger.Log("debug", fmt.Sprintf("ELB '%s' already exists, reusing", lb.Name))
 	}
 
-	// create DNS record for LB
+	// create DNS records for LB
```
Comment (xh3b4sd) [inline service/create/loadbalancer.go]: Not your fault but I would like to see comments being sentences in general.

File: service/create/loadbalancer.go [hunk 3]
```diff
@@ -23,12 +23,14 @@ type LoadBalancerInput struct {
...
-		InstancePort:     input.Cluster.Spec.Cluster.Kubernetes.API.SecurePort,
-		LoadBalancerPort: input.Cluster.Spec.Cluster.Kubernetes.API.SecurePort,
-		Client:           input.Clients.ELB,
+		Name:          input.Cluster.Spec.Cluster.Cluster.ID,
+		SecurityGroup: input.SecurityGroupID,
+		SubnetID:      input.SubnetID,
+		PortsToOpen: []int{
+			input.Cluster.Spec.Cluster.Kubernetes.API.SecurePort,
+			input.Cluster.Spec.Cluster.Etcd.Port,
+		},
```
Comment (xh3b4sd) [inline service/create/loadbalancer.go]: I like that very much. It reads to clearly with the decent structure of the TPR.

File: service/create/loadbalancer.go [hunk 4]
```diff
@@ -65,18 +67,31 @@ func (s *Service) createLoadBalancer(input LoadBalancerInput) error {
...
+		Resource:     lb,
+		Domain:       input.Cluster.Spec.Cluster.Etcd.Domain,
+		HostedZoneID: hz.ID(),
+	}
+
+	if err := etcdRecordSet.CreateOrFail(); err != nil {
+		return microerror.MaskAny(fmt.Errorf("error registering DNS '%s'", errgo.Details(err)))
+	}
+
+	s.logger.Log("debug", fmt.Sprintf("created or reused DNS record for etcd"))
```
Comment (xh3b4sd) [inline service/create/loadbalancer.go]: When there is no variable to substitute there is no reason to use `fmt`. The same applies to the logging below.

File: resources/aws/elb.go [hunk 5]
```diff
@@ -47,16 +46,21 @@ func (lb *ELB) CreateOrFail() error {
 		return microerror.MaskAny(clientNotInitializedError)
 	}
 
+	var listeners []*elb.Listener
+	for _, portToOpen := range lb.PortsToOpen {
+		listener := &elb.Listener{
+			InstancePort:     aws.Int64(int64(lb.PortsToOpen[portToOpen])),
+			LoadBalancerPort: aws.Int64(int64(lb.PortsToOpen[portToOpen])),
+			// TCP because we want to do SSL passthrough, not termination
```
Comment (xh3b4sd) [inline resources/aws/elb.go]: Not your fault but I would like to see comments being sentences in general.

File: examples/cluster.yml [hunk 6]
```diff
@@ -16,6 +16,7 @@ spec:
     etcd:
       domain: "k8s.cluster.giantswarm.io"
       prefix: "example-cluster"
+      port: 2379
```
Comment (xh3b4sd) [inline examples/cluster.yml]: So what is the idea conceptually? Exposing etcd on its well known standard port via domain using TLS? I remember we defined two ports being open for etcd. Will one be additionally open internally as well? Would like to have a clear picture to get that straight.

File: service/create/loadbalancer.go (hunk 1 above)
Comment (rossf7) [inline service/create/loadbalancer.go]: I agree on the typed error but its happening in a lot of places. I'd prefer to raise an issue and tidy this up later.

File: examples/cluster.yml (hunk 6 above)
Comment (rossf7) [inline examples/cluster.yml]: I've added more detail to the PR and why only port 2379 is being opened.

**Comment by asymmetric** [comment]
PTALA. I also added some more context/questions in the PR description, PTAL.

File: service/create/loadbalancer.go (hunk 1 above)
Comment (asymmetric) [inline service/create/loadbalancer.go]: There is [an issue](https://github.com/giantswarm/aws-operator/issues/123) btw.

File: service/create/service.go [hunk 7]
```diff
@@ -644,8 +645,9 @@ func (s *Service) Boot() {
 					}
 					if err := kmsKey.Delete(); err != nil {
 						s.logger.Log("error", errgo.Details(err))
```
Comment (xh3b4sd) [inline service/create/service.go]: What happens in this error case? Should we go ahead? Should we retry?

Comment (asymmetric) [inline service/create/service.go]: At the moment we go ahead, same as we do for other errors on deletion.

* If the resource is not there, we should just go ahead
* If there was a temporary error (say, a network error), we could retry

I'd say this is a more general issue we should tackle in the scheme of making the operator more resilient. There was talk of having a focus on this after we get the basic functionality in place.

Comment (xh3b4sd) [inline service/create/service.go]: Ok cool.

**Pull request merged**